### PR TITLE
chore: add type definition for jest mock

### PR DESCRIPTION
**Pull request by lafest**
## Summary

According to [with-jest-setup-file jest integration](https://react-native-async-storage.github.io/async-storage/docs/advanced/jest#with-jest-setup-file), I tried to test `app.test.tsx` file using [ts-jest](https://kulshekhar.github.io/ts-jest/) and followed 
```ts
import mockAsyncStorage from '@react-native-async-storage/async-storage/jest/async-storage-mock';
```
throws ts(7016) error.
스크린샷 2021-11-09 오후 2 58 29

## Test Plan

If you use this change with ts-jest, the error disappears.

### CODE REVIEW

**Comment by lafest** [comment]
@Krizzu please review this PR. It seems that you're the maintainer.

**Comment by Krizzu** [comment]
Sure thing, looks fine

I just wonder if we can reduce the duplication and do something like this

```
// CREDITS: This types are based on the original work made by all the people who contributed to @types/react-native

interface AsyncStorage {
  /**
   * Fetches key and passes the result to callback, along with an Error if there is any.
   */
  getItem(key: string, callback?: (error?: Error, result?: string) => void): Promise<string | null>;

  /**
   * Sets value for key and calls callback on completion, along with an Error if there is any
   */
  setItem(key: string, value: string, callback?: (error?: Error) => void): Promise<void>;

  removeItem(key: string, callback?: (error?: Error) => void): Promise<void>;

  /**
   * Merges existing value with input value, assuming they are stringified json. Returns a Promise object.
   * Not supported by all native implementation
   */
  mergeItem(key: string, value: string, callback?: (error?: Error) => void): Promise<void>;

  /**
   * Erases all AsyncStorage for all clients, libraries, etc. You probably don't want to call this.
   * Use removeItem or multiRemove to clear only your own keys instead.
   */
  clear(callback?: (error?: Error) => void): Promise<void>;

  /**
   * Gets all keys known to the app, for all callers, libraries, etc
   */
  getAllKeys(callback?: (error?: Error, keys?: string[]) => void): Promise<string[]>;

  /**
   * multiGet invokes callback with an array of key-value pair arrays that matches the input format of multiSet
   */
  multiGet(
    keys: string[],
    callback?: (errors?: Error[], result?: [string, string | null][]) => void
  ): Promise<[string, string | null][]>;

  /**
   * multiSet and multiMerge take arrays of key-value array pairs that match the output of multiGet,
   *
   * multiSet([['k1', 'val1'], ['k2', 'val2']], cb);
   */
  multiSet(keyValuePairs: string[][], callback?: (errors?: Error[]) => void): Promise<void>;

  /**
   * Delete all the keys in the keys array.
   */
  multiRemove(keys: string[], callback?: (errors?: Error[]) => void): Promise<void>;

  /**
   * Merges existing values with input values, assuming they are stringified json.
   * Returns a Promise object.
   *
   * Not supported by all native implementations.
   */
  multiMerge(keyValuePairs: string[][], callback?: (errors?: Error[]) => void): Promise<void>;
}

type AsyncStorageHook = {
  getItem(callback?: (error?: Error, result?: string) => void): Promise<string | null>;
  setItem(value: string, callback?: (error?: Error) => void): Promise<void>;
  mergeItem(value: string, callback?: (error?: Error) => void): Promise<void>;
  removeItem(callback?: (error?: Error) => void): Promise<void>;
}

declare module '@react-native-async-storage/async-storage' {
  /**
   * AsyncStorage is a simple, unencrypted, asynchronous, persistent, key-value storage
   * system that is global to the app.  It should be used instead of LocalStorage.
   *
   * It is recommended that you use an abstraction on top of `AsyncStorage`
   * instead of `AsyncStorage` directly for anything more than light usage since
   * it operates globally.
   *
   * On iOS, `AsyncStorage` is backed by native code that stores small values in a
   * serialized dictionary and larger values in separate files. On Android,
   * `AsyncStorage` will use either [RocksDB](http://rocksdb.org/) or SQLite
   * based on what is available.
   *
   * @see https://react-native-async-storage.github.io/async-storage/docs/api
   */
  export function useAsyncStorage(key: string): AsyncStorageHook
  const AsyncStorageLib: AsyncStorage;
  export default AsyncStorageLib;
}

declare module '@react-native-async-storage/async-storage/jest/async-storage-mock' {
  export function useAsyncStorage(key: string): AsyncStorageHook
  const AsyncStorageLib: AsyncStorage;
  export default AsyncStorageLib;
}
```

**Comment by lafest** [comment]
@Krizzu Thanks for review!
I modified PR with based on your review.

It seems like some internal error in Android testing, but I have no permission for rerunning CI tests. Can you rerun that CI test?

**Comment by Krizzu** [comment]
@lafest Hey, please rebase your code, we merge a fix for this yesterday

**Comment by lafest** [comment]
@Krizzu I rebased this repository's master branch to mine. I'm not sure I did right thing, because this is my first rebase. Did I do right?

**Comment by Krizzu** [comment]
It's more like a merge, than a rebase, but it's fine.

@tido64  mind having a look at the types? Need another pair of eyes on this 🙏

**Review by tido64 — COMMENTED** [review]
Except for the merge/rebase weirdness, it looks fine to me.

File: types/index.d.ts
Comment (lafest) [reply types/index.d.ts:87]: It worked well in mine. Because this is default export, it does not matter. If you means that this seems like wrong naming convention, I'll fix it.

Comment (lafest) [reply types/index.d.ts:87]: It worked well for mine. Because this is default export, naming did not matter.
If you means it seems like wrong naming convention, I'll fix it.

Comment (tido64) [reply types/index.d.ts:87]: I was just wondering if we need this line at all: https://github.com/react-native-async-storage/async-storage/blob/ed50c7721978404c1532ea077f93f812745f2a05/types/index.d.ts#L86

Why can't it just be:
```ts
export default AsyncStorage;
```

Comment (lafest) [reply types/index.d.ts:87]: Because AsyncStorage itself is type, if we export it directly, It makes error like image.

스크린샷 2021-11-24 오후 5 03 13

Comment (tido64) [reply types/index.d.ts:87]: Ah, that makes sense. Thanks for educating me 😄

**Review by tido64 — APPROVED** [review]

**Pull request merged**
